### PR TITLE
OpenTelemetry Project Security Self-Assessment

### PR DESCRIPTION
**Pull request by PrathamGupta**
Created and added first draft of the [OpenTelemetry](https://opentelemetry.io) Project Security Self-Assessment.
Please feel free to share any thoughts on the self-assessment.

### CODE REVIEW

**Comment by eddie-knight** [comment]
Hi there! I'm just getting started looking at your pull request, and I noticed the DCO check is failing.

You can look at the checks section of the PR (I believe it should always be below the last comment) and look for a red X highlighting the failed check. In this case, you can click `Details` for more information about how to get that check passing.
Screenshot 2023-12-08 at 8 35 18 AM

File: assessments/projects/open-telemetry/self-assessment-template.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
+# Self-assessment
```
Comment (eddie-knight) [inline assessments/projects/open-telemetry/self-assessment-template.md:1]:
```suggestion
# Open Telemetry Self-assessment
```

**Review by ragashreeshekar — CHANGES_REQUESTED** [review]
Thanks for the PR @PrathamGupta and team, appreciate the efforts.
I have completed first pass of review. Please feel free to reach out here or on slack for any questions and clarifications.

Along with addressing the comments, kindly update the PR branch with the latest content in the repo as this branch is out-of-date with the base branch.

File: assessments/projects/open-telemetry/self-assessment-template.md
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment-template.md:1]: We have updated our PR with removing all the template content and tailored it for these needs. We have also corrected all the checks and will keep the PR updated for these checks in the future.

**Comment by PrathamGupta** [comment]
> Thanks for the PR @PrathamGupta and team, appreciate the efforts. I have completed first pass of review. Please feel free to reach out here or on slack for any questions and clarifications.
> 
> Along with addressing the comments, kindly update the PR branch with the latest content in the repo as this branch is out-of-date with the base branch.

Thank you @ragashreeshekar for reviewing our work! We have updated the PR branch with the latest content in the repo and will be updated regularly till the PR has been reviewed completely.

File: assessments/projects/open-telemetry/self-assessment.md [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+
+* Data limitations: OpenTelemetry can help you access high-quality telemetry data, but data storage and visualization still require other tools, and code analysis and application security are unavailable.
+
+**Security**
+
+
+
+* Use API tokens and TLS network security encryption to encrypt sensitive information to prevent third parties from querying OpenTelementry telemetry data.
+* Prevent external access to internal resources. When running Open Telemetry, there should be no privileged access.
+* The agents must authenticate themselves to the backend systems where the data is sent.
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md]: We have updated the same for Non Goals and we will also try to update the goals and add more details.

File: assessments/projects/open-telemetry/self-assessment.md [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+**General**
+
+
+
+* Custom telemetry: Set sensible defaults but allow customization.
+* Own your data: Not tied to proprietary data formats or tools.
+* High performance: Unexpected interference effects do not exist in the host application.
+* Simplicity: It is only necessary to learn a single set of APIs and conventions to capture traces and metrics, simplifying the application instrumentation process.
+
+**Security**
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md:186]: We have updated the goals of the document and increased the scope in this section.

Thanks for the feedback

File: assessments/projects/open-telemetry/self-assessment.md [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+
+
+* **Prometheus**: Both Prometheus and OpenTelemetry can collect metrics across applications and services. However, Prometheus focuses more on the pull model, and OpenTelemetry includes both push and pull models. In addition, OpenTelemetry does not have built-in alarms, but Prometheus does. OpenTelemetry combining built-in alerts in Prometheus helps create an observability ecosystem and improve stability.
+* **Grafana**: OpenTelemetry and Grafana complement each other, and the visualization in Grafana's dashboard can better help analyze various types of data collected from OpenTelemetry. In addition, Grafana supports enable interoperability from OpenTelemetry by calling on SDKs and application observability standards, among others, integrating telemetry into a unified open-source monitoring backend.
+* **OpenTelemetry**: Same
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md]: Noted and changed.

Thanks for the feedback!

File: assessments/projects/open-telemetry/self-assessment.md [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+* **Jaeger**: Jaeger is a distributed tracing system that can be started and run in a local environment. Users can first use OpenTelemetry to detect various indicators of the application and then send the tracking data to Jaeger. The advantage of using Jaeger is that it makes finding the source of latency and lack of concurrency easier. All in all, Jaeger is more focused, while OpenTelemetry is a broader framework.
+* **Cilium**: Not applicable
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md]: The things which are not applicable are removed. 

Thanks for feedback!

File: assessments/projects/open-telemetry/self-assessment.md [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+
+
+### Known Issues Over Time
+
+The Enhancements Telemetry security-related issues can be searched using the keywords "Security type:pr," and the fixes can be searched using the keywords "fix type:pr." it can be queried using[ https://github.com/open-telemetry/opentelemetry.io/issues](https://github.com/open-telemetry/opentelemetry.io/issues).
+
+Consider security when making pull requests, and need to consider the possibility of exposing sensitive information.
+
+
+### CII Best Practices
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md]: We have added this as a recommendation for the project!

**Comment by PrathamGupta** [comment]
@ragashreeshekar We have updated the PR with updates from your suggestions! Please let us know your thoughts about the assessment now! 

Thanks!

**Comment by PrathamGupta** [comment]
@torinvdb We are currently in the process of making the changes you have suggested and going through your feedback

File: assessments/projects/open-telemetry/self-assessment.md [hunk 7]
```diff
@@ -0,0 +1,340 @@
+# OpenTelemetry Security Self Assessment
+
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md:2]: We have added this disclaimer again! We removed it on our assumption that this was not necessary, but we know understand the use of this disclaimer!

Thanks for the feedback

File: assessments/projects/open-telemetry/self-assessment.md [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+### Communication Channels
+* Internal
+  * The OpenTelemetry team mostly uses platforms like GitHub, Slack, or email lists for internal communications within the teams.
+* Inbound
+  * Users and contributors to the OpenTelemetry project can communicate with the OpenTelemetry team via GitHub issues, mailing lists, CNCF Slack channels and through Discord Communications as well.
+* Outbound
+  * The updates and announcements from OpenTelemetry are made through OpenTelemetry Blog, GitHub, CNCF mailing lists, and social media channels.
+
+### Ecosystem
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md:292]: We have added the reference to our assessment as well

File: assessments/projects/open-telemetry/self-assessment.md [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+
+
+## Overview
+
+OpenTelemetry is an observability framework and toolkit that helps developers understand application behavior by instrumenting, generating, and exporting telemetry data for applications using OpenTelemetry. This includes open-source backends such as Jaeger and Prometheus, in addition to commercial offerings.
+
+
+### Background
+
+OpenTelemetry is an Observability framework and toolkit designed to create and manage telemetry data such as traces, metrics, and logs. Opentelemetry provides a standard protocol for dealing with telemetry data, language SDKs, a collector that receives, processes, and exports telemetry data, as well as other documentation and tools for telemetry.
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md:109]: We have changed this in our assessment

File: assessments/projects/open-telemetry/self-assessment.md [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+
+**OpenTelemetry SDK / OTLP**
+
+OpenTelemetry provides various SDKs for C++, .NET, Erlang / Elixir, Go, Java, JavaScript / TypeScript, PHP, Python, Ruby, Rust, Swift. These SDKs provide instrumentation for a system, and export raw traces, metrics, and logs.
+
+This instrumentation can be automatic, in which case, they export relevant telemetry to the OpenTelemetry Collector with minimal configuration from the user. Alternatively, they can be manual. In this case, the user manually configures the system using the SDK and OpenTelemetry API to provide the relevant data to the Collector.
+
+Furthermore, a third party application that produces data in the OTLP format may be used to send telemetry to the collector.
+
+As these components are the primary source of incoming data, they also serve as the primary attack surface. A hypothetical attacker may be able to deliver a malicious payload by simply performing an action that is logged. They can use this to either attack the SDK and try and gain control of the system directly, or to attack components further down the pipeline, such as the Collector.
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md]: We have reviewed and changed this as well.

File: assessments/projects/open-telemetry/self-assessment.md [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+
+**OpenTelemetry Collector**
+
+The OpenTelemetry Collector is a vendor-agnostic implementation of how to receive, process and export telemetry data. It consists of multiple components: Receivers, Processors, Exporters, and Extensions. It serves as middleware between the raw telemetry feeds and telemetry frontends. It receives data via the Receiver using the OpenTelemetry Protocol. The collector then processes the data, handling jobs such as retries, batching, and encryption. It then exports the data to an Observability frontend, like Jaeger or Prometheus for analysis.
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md]: We shall change this in other sections as well

File: assessments/projects/open-telemetry/self-assessment.md [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+* Prevent external access to internal resources. When running Open Telemetry, there should be no privileged access.
+* The agents must authenticate themselves to the backend systems where the data is sent.
+
+
+## Self-Assessment Use
+
+Our team comprises graduate students from different professional backgrounds, and each member is rich in insight. We have come together with a shared commitment to create this self-assessment to analyze the overall security of OpenTelemetry. This document mainly helps users get a preliminary understanding of OpenTelemetry’s security architecture and security development practices. In addition, it allows users to have a deeper understanding of the security functions of OpenTelemetry, helping users better develop, use, and telemetry the data they need.
+
+All in all, this document helps explain this CNCF open source project, pave the way for the project to seek graduation, provide clear guidance during the incubation process, and bring a more transparent and clear security audit to the overall project.
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md:215]: Thank you for the insights regarding this and we have changed this in our assessment as well

File: assessments/projects/open-telemetry/self-assessment.md [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+### CII Best Practices
+
+Not applicable
+
+
+### Case Studies
+
+We've seen users of OpenTelemetry use the following way:
+
+* OpenTelemetry makes generating, collecting, and exporting data from applications easy. We can use OpenTelemetry for system monitoring and then use a visual monitoring system to integrate the data sets generated by OpenTelemetry. For example, use AWS to build a simple microservice. After completing this configuration, use OpenTelemetry in the service to parse and track some contextual information. After the data collection, use Jaeger to connect to OpenTelemetry to obtain a more beautiful visual interface. When we try to access Jaeger, we can see some exception stack information, which helps us analyze online exceptions and understand the operating response of the system.
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md:354]: We have added one example of use case in Kubernetes from the specifications mentioned on the websites. Let us know if we should add for more of the integrations

File: assessments/projects/open-telemetry/self-assessment.md [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+
+Once processed, the data is exported to various monitoring and analysis tools (like Prometheus, Jaeger, etc.). The export process includes secure API calls or data transmission methods. The receiving systems often authenticate the incoming data to ensure its validity.
+
+**Analysis and Visualization**
+
+The data is analyzed, and insights are presented through dashboards or alerts. Access to these insights is typically controlled through user authentication and authorization, ensuring that only authorized users can view or manipulate sensitive telemetry data.
+
+**Alerting and Incident Response**
+
+In case of anomalies or incidents detected through telemetry data, alerts are generated. The alerting mechanism is secured to prevent false alerts and ensure that alerts reach the correct recipients. Incident response protocols are in place to handle any security incidents reported through these alerts.
```
[the remark this answers is not in the document]
Comment (PrathamGupta) [reply assessments/projects/open-telemetry/self-assessment.md:179]: We have ammended the same and answered some of the questions to our best ability.

File: assessments/projects/open-telemetry/self-assessment.md [hunk 15]
```diff
@@ -0,0 +1,376 @@
+# OpenTelemetry Security Self Assessment
+
+The Self-assessment is the initial document for projects to begin thinking about the security of the project, determining gaps in their security, and preparing any security documentation for their users. This document is ideal for projects currently in the CNCF sandbox as well as projects that are looking to receive a joint assessment and currently in CNCF **Incubation**.
```
Comment (torinvdb) [inline assessments/projects/open-telemetry/self-assessment.md:3]:
```suggestion
# OpenTelemetry Security Self Assessment

This assessment was created by community members as part of the [Security Pals](https://github.com/cncf/tag-security/issues/1102) process and is currently pending changes from the maintainer team.

The Self-assessment is the initial document for projects to begin thinking about the security of the project, determining gaps in their security, and preparing any security documentation for their users. This document is ideal for projects currently in the CNCF sandbox as well as projects that are looking to receive a joint assessment and currently in CNCF **Incubation**.
```
@ragashreeshekar

**Review by ragashreeshekar — APPROVED** [review]

File: assessments/projects/open-telemetry/self-assessment.md [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+* **Grafana**: OpenTelemetry and Grafana complement each other, and the visualization in Grafana's dashboard can better help analyze various types of data collected from OpenTelemetry. In addition, Grafana supports enable interoperability from OpenTelemetry by calling on SDKs and application observability standards, among others, integrating telemetry into a unified open-source monitoring backend.
+* **Jaeger**: Jaeger is a distributed tracing system that can be started and run in a local environment. Users can first use OpenTelemetry to detect various indicators of the application and then send the tracking data to Jaeger. The advantage of using Jaeger is that it makes finding the source of latency and lack of concurrency easier. All in all, Jaeger is more focused, while OpenTelemetry is a broader framework.
+
+[comment]: # (The Final Draft of Self Assessment has been completed on Dec 14, 2023: 5:40 EST)
```
Comment (JustinCappos) [inline assessments/projects/open-telemetry/self-assessment.md:352]:
```suggestion
```

**Review by JustinCappos — APPROVED** [review]
LGTM.  Ready to merge!

**Pull request merged**